extensions/firmware/org.eclipse.smarthome.firmware.filesystem/src/main/java/org/eclipse/smarthome/firmware/filesystem/FileSystemFirmwareProvider.java

```diff
@@ -0,0 +1,291 @@
+
+        InputStream firmwareImage = getFirmwareImage(firmwareUID, rb.getString("image"));
+
+        return new Firmware.Builder(firmwareUID).withDescription(getResourceAsString(rb, "description"))
+                .withModel(getResourceAsString(rb, "model"))
+                .withVendor(getResourceAsString(rb, "vendor")).withChangelog(getResourceAsString(rb, "changelog"))
+                .withPrerequisiteVersion(getResourceAsString(rb, "prerequisiteVersion"))
+                .withOnlineChangelog(new URL(getResourceAsString(rb, "onlineChangelog"))).withContent(firmwareImage)
+                .build();
+    }
+
+    private InputStream getFirmwareImage(FirmwareUID firmwareUID, String filename) throws IOException {
+        ThingTypeUID thingTypeUID = firmwareUID.getThingTypeUID();
+        Path imageFile = Paths.get(rootPath, thingTypeUID.getBindingId(), thingTypeUID.getId(), filename);
+    
+        if (Files.notExists(imageFile)) {
+            throw new IOException("File " + imageFile.getFileName().toString() + " cannot be found");
+        }
+    
+        return FileUtils.openInputStream(imageFile.toFile());
+    }
+
+    private String getResourceAsString(ResourceBundle rb, String key) {
+        return rb.containsKey(key) ? rb.getString(key) : null;
+    }
+
+    @Override
+    public Set<Firmware> getFirmwares(ThingTypeUID thingTypeUID) {
+        return getFirmwares(thingTypeUID, Locale.getDefault());
+    }
+
+    @Override
+    public Set<Firmware> getFirmwares(final ThingTypeUID thingTypeUID, Locale locale) {
+        Preconditions.checkArgument(thingTypeUID != null);
+        if (locale == null) {
+            locale = Locale.getDefault();
+        }
+
+        final Locale myLocale = locale;
+
+        Path thingTypePath = Paths.get(rootPath, thingTypeUID.getBindingId(), thingTypeUID.getId());
+        final Set<Firmware> firmwareSet = new HashSet<>();
+
+        try {
+            Files.walkFileTree(thingTypePath, new SimpleFileVisitor<Path>() {
+                @Override
+                public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
+                    if (PROPERTY_FILE_MATCHER.matches(file)) {
+                        FirmwareUID firmwareUID = getFirmwareUIDFromFileName(thingTypeUID,
+                                getBaseName(file.getFileName().toString()));
+                        if (firmwareUID != null) {
+                            Firmware firmware = getFirmware(firmwareUID, myLocale);
+                            if (firmware != null) {
+                                firmwareSet.add(firmware);
+                            }
+                        }
+                    }
+                    return FileVisitResult.CONTINUE;
+                }
+            });
+        } catch (IOException e) {
+            logger.error("Error searching firmwares for thing type " + thingTypeUID, e);
```

Comment:
We should not log each IOException as error. E.g. if you invoke the operation for a thing type UID for which no firmware exists then a NoSuchFileException is thrown. I think we should only print a debug message for this exception without the stack trace.